File: html_minifier.rb
# html_minifier.rb
class HTMLMinifier
  def initialize(remove_comments: true, preserve_whitespace: false, remove_quotes: false)
    @remove_comments = remove_comments
    @preserve_whitespace = preserve_whitespace
    @remove_quotes = remove_quotes
  end

  def minify(html)
    html = html.gsub(/<!--.*?-->/m, '') if @remove_comments

    lines = html.lines.map(&:strip).reject(&:empty?)
    html = lines.join

    unless @preserve_whitespace
      html.gsub!(/>\s+</, '><')
      html.gsub!(/ {2,}/, ' ')
    end

    if @remove_quotes
      html.gsub!(/=\s*"([^"]*)"/, '=\1')
      html.gsub!(/=\s*'([^']*)'/, '=\1')
    end

    html.gsub!(/\n\s*\n/, "\n")
    html.strip
  end
end

def main
  input_file = nil
  output_file = nil
  no_comments = false
  preserve_whitespace = false
  remove_quotes = false

  args = ARGV
  until args.empty?
    arg = args.shift
    case arg
    when '--no-comments' then no_comments = true
    when '--preserve-whitespace' then preserve_whitespace = true
    when '--remove-quotes' then remove_quotes = true
    when '-o' then output_file = args.shift
    else input_file = arg
    end
  end

  if input_file
    content = File.read(input_file)
  else
    content = $stdin.read
  end

  minifier = HTMLMinifier.new(
    remove_comments: !no_comments,
    preserve_whitespace: preserve_whitespace,
    remove_quotes: remove_quotes
  )
  minified = minifier.minify(content)

  if output_file
    File.write(output_file, minified)
  else
    puts minified
  end
end

main if __FILE__ == $0

File: html_minifier.py

---

# 💻 Code Implementations

## 1. Python (`html_minifier.py`)

```python
# html_minifier.py
import re
import sys
import argparse

class HTMLMinifier:
    def __init__(self, remove_comments=True, preserve_whitespace=False, remove_quotes=False):
        self.remove_comments = remove_comments
        self.preserve_whitespace = preserve_whitespace
        self.remove_quotes = remove_quotes

    def minify(self, html):
        # Remove comments
        if self.remove_comments:
            html = re.sub(r'<!--.*?-->', '', html, flags=re.DOTALL)

        # Remove leading/trailing whitespace on each line
        lines = html.splitlines()
        html = ' '.join(line.strip() for line in lines)

        # Collapse multiple spaces
        if not self.preserve_whitespace:
            # Remove spaces between tags (but keep at least one space if needed)
            html = re.sub(r'>\s+<', '><', html)
            # Collapse multiple spaces inside text (but keep single spaces)
            html = re.sub(r' {2,}', ' ', html)
            # Remove spaces around attributes? Skip for simplicity.

        # Remove quotes from simple attributes (optional)
        if self.remove_quotes:
            html = re.sub(r'=\s*"([^"]*)"', r'=\1', html)
            html = re.sub(r"=\s*'([^']*)'", r'=\1', html)

        # Remove empty lines
        html = re.sub(r'\n\s*\n', '\n', html)
        return html.strip()

def main():
    parser = argparse.ArgumentParser(description='HTML Minifier')
    parser.add_argument('input', nargs='?', help='Input HTML file (or stdin)')
    parser.add_argument('-o', '--output', help='Output file (or stdout)')
    parser.add_argument('--no-comments', action='store_true', help='Keep comments')
    parser.add_argument('--preserve-whitespace', action='store_true', help='Preserve whitespace inside text')
    parser.add_argument('--remove-quotes', action='store_true', help='Remove quotes from simple attributes')
    args = parser.parse_args()

    if args.input:
        with open(args.input, 'r', encoding='utf-8') as f:
            content = f.read()
    else:
        content = sys.stdin.read()

    minifier = HTMLMinifier(
        remove_comments=not args.no_comments,
        preserve_whitespace=args.preserve_whitespace,
        remove_quotes=args.remove_quotes
    )
    minified = minifier.minify(content)

    if args.output:
        with open(args.output, 'w', encoding='utf-8') as f:
            f.write(minified)
    else:
        print(minified)

if __name__ == '__main__':
    main()
